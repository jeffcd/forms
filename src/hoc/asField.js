import React from 'react'
import get from 'lodash/get'
import FormContext from '../FormContext'
import VisibilityContext from '../VisibilityContext'
import ListContext from '../ListContext'
import validateField, { validatorFunctions } from '../validateField'
import messages from '../messages'
const validators = validatorFunctions

const getValidators = props => {
  const fieldValidators = {}
  Object.keys(props).forEach(name => {
    if (validators[name]) {
      fieldValidators[name] = props[name]
    }
  })
  return fieldValidators
}

const errorMessages = messages

const getErrorMessages = ({ props, field }) => {
  return field.errors.map(error => {
    let message = errorMessages[error]
    if (props.messages && props.messages[error]) {
      message = props.messages[error]
    }
    return message
      .replace('%name%', props.label)
      .replace('%param%', props[error])
  })
}

const getInitialValue = (Field, { minLength = 0 }) => {
  if (Field.type === 'list') {
    const size = parseInt(minLength)
    return Array(size)
  }

  return ''
}

const asFieldHoc = Field => {
  return function asFieldHoc({ ...rest }) {
    return (
      <ListContext.Consumer>
        {listInfo => {
          const { name, initialValue, convertTo } = rest
          const fullName = listInfo
            ? `${listInfo.name}.value[${listInfo.i}].${name}`
            : name

          const listProps = {}
          if (listInfo) {
            const { i, name } = listInfo
            const fullName = `${name}.value[${i}].${rest.name}`
            const id = `${name}:${i}::${rest.name}:`
            listProps.id = id
            if (rest.name) {
              listProps.name = fullName
            }
          }

          return (
            <FormContext.Consumer>
              {form => {
                //useEffect(() => {
                //}, [form.setFormField])

                const field = get(form.fields, fullName)
                if (!field) {
                  const field = {
                    name,
                    value: initialValue || getInitialValue(Field, rest),
                    validators: getValidators(rest),
                    convertTo,
                    errors: []
                  }
                  const isUser = false
                  form.setFormField(field, fullName, isUser)

                  return null
                }

                return (
                  <VisibilityContext.Consumer>
                    {visibility => {
                      if (visibility.isVisible === false) {
                        if (field.isVisible !== false) {
                          const isUser = false
                          field.isVisible = false
                          form.setFormField({ ...field }, isUser)
                        }
                        return null
                      }

                      const handleChange = e => {
                        const field = get(form.fields, fullName)
                        const newField = {
                          ...field,
                          value: e.target.value
                        }
                        const errors = validateField(newField)
                        newField.errors = errors
                        form.setFormField(newField, fullName)
                      }
                      return (
                        <Field
                          onChange={handleChange}
                          value={field.value}
                          errors={getErrorMessages({ props: rest, field })}
                          {...rest}
                          {...listProps}
                        />
                      )
                    }}
                  </VisibilityContext.Consumer>
                )
              }}
            </FormContext.Consumer>
          )
        }}
      </ListContext.Consumer>
    )
  }
}

export default asFieldHoc
