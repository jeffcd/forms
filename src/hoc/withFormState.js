import React from 'react'
import FormContext from '../FormContext'
import stateStrs from '../formStates'

const withFormStateHoc = Observer => {
  return function withFormStateHoc({ ...rest }) {
    return (
      <FormContext.Consumer>
        {form => {
          if (!form) {
            return null
          }
          return <Observer form={form} stateStrs={stateStrs} {...rest} />
        }}
      </FormContext.Consumer>
    )
  }
}

export default withFormStateHoc
