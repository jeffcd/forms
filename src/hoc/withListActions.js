import React from 'react'
import get from 'lodash/get'
import FormContext from '../FormContext'
import ListContext from '../ListContext'
import ScopeContext from '../ScopeContext'

const withListActionsHoc = Component => {
  return function withListActionsHoc({ ...rest }) {
    return (
      <FormContext.Consumer>
        {form => {
          return (
            <ScopeContext.Consumer>
              {scope => {
                return (
                  <ListContext.Consumer>
                    {listInfo => {
                      const listProps = {}
                      const to = rest.to
                      const from = rest.from
                      if (listInfo) {
                        const { name, i } = listInfo
                        if (to) {
                          listProps.to = `${name}.value[${i}].${to}`

                          const field = get(form.fields, listProps.to)
                          const listInfo = field.validators
                          const length = field.value.length
                          if (
                            to &&
                            listInfo.maxLength &&
                            length >= listInfo.maxLength
                          ) {
                            return null
                          }
                        }
                        if (from) {
                          listProps.from = `${name}`
                          listProps.index = i
                          if (
                            listInfo.minLength &&
                            listInfo.length <= listInfo.minLength
                          ) {
                            return null
                          }
                        }
                      } else {
                        const fullTo = (scope ? `${scope}.` : '') + to
                        const field = get(form.fields, fullTo)
                        const listInfo = field.validators
                        const length = field.value.length
                        listProps.to = fullTo
                        if (
                          to &&
                          listInfo.maxLength &&
                          length >= listInfo.maxLength
                        ) {
                          return null
                        }
                      }
                      return (
                        <Component
                          listActions={form.listActions}
                          {...rest}
                          {...listProps}
                        />
                      )
                    }}
                  </ListContext.Consumer>
                )
              }}
            </ScopeContext.Consumer>
          )
        }}
      </FormContext.Consumer>
    )
  }
}

export default withListActionsHoc
