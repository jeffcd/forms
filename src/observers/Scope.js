import React, { useState } from 'react'
import ScopeContext from '../ScopeContext'

const Scope = ({ children, path = '' }) => {
  const [fullPath, setFullPath] = useState(false)
  return (
    <ScopeContext.Consumer>
      {scope => {
        if (!fullPath) {
          if (scope) {
            setFullPath(`${scope}.${path}`)
          } else {
            setFullPath(path)
          }
          return null
        }
        return (
          <ScopeContext.Provider value={fullPath}>
            {children}
          </ScopeContext.Provider>
        )
      }}
    </ScopeContext.Consumer>
  )
}

export default Scope
