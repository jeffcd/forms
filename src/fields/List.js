import React from 'react'
import ListContext from '../ListContext'
import asField from '../hoc/asField'

const List = ({
  children,
  name,
  value,
  minLength = 0,
  maxLength = 1000000
}) => {
  const arr = Array(value.length).fill(1)
  const childrenArr = Array.isArray(children) ? children : [children]
  return (
    <>
      {arr.map((a, i) => {
        const listInfo = {
          name,
          i,
          minLength: parseInt(minLength),
          maxLength: parseInt(maxLength),
          length: value.length
        }
        return (
          <React.Fragment key={i}>
            <ListContext.Provider value={listInfo}>
              {childrenArr.map((child, j) => {
                if (!child.props) {
                  return child
                }
                const Clone = React.cloneElement(child)
                return <React.Fragment key={j}>{Clone}</React.Fragment>
              })}
            </ListContext.Provider>
          </React.Fragment>
        )
      })}
    </>
  )
}

List.type = 'list'

export default asField(List)
