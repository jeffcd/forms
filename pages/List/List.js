import React from 'react'
import List from '../../src/fields/List'
import SimpleStates from '../../src/observers/SimpleStates'
import asField from '../../src/hoc/asField'
import withFormState from '../../src/hoc/withFormState'
import withListActions from '../../src/hoc/withListActions'
import VisibilityGroup from '../../src/observers/VisibilityGroup'
import Form from '../../src'

const Input = asField(
  ({
    onChange,
    onBlur = () => {},
    name,
    id,
    value,
    errors,
    label,
    helperText,
    required = false
  }) => {
    return (
      <div>
        <label htmlFor={name}>{label}</label>
        <br />
        <input
          id={`input-${id}`}
          name={name}
          value={value}
          onChange={e => onChange(e)}
          onBlur={e => onBlur(e)}
          label={label}
          required={required}
        />
        <span>{errors.length ? errors[0] : helperText}</span>
      </div>
    )
  }
)

const Button = withFormState(({ text, type, onClick, form }) => {
  const disabled = form.state === 'submit'

  return (
    <button type={type} disabled={disabled} onClick={onClick}>
      {text}
    </button>
  )
})

const AddListItem = withListActions(({ text, to, listActions }) => {
  return (
    <Button
      text={text}
      type="button"
      onClick={() => listActions.addItemToList(to)}
    />
  )
})

const RemoveListItem = withListActions(({ text, from, index, listActions }) => {
  return (
    <Button
      text={text}
      type="button"
      onClick={() => listActions.removeItemFromList(from, index)}
    />
  )
})

const nameMessages = {
  match: 'Your name must begin with Jeff.'
}

const stateStrs = {
  error_server: 'There was a server error.'
}

const PList = () => {
  const handleSubmit = async (e, body, actions) => {
    e.preventDefault()
    console.log(body)
    console.log(actions)
    // This is just for example...
    const result = await new Promise(res => {
      setTimeout(() => {
        // actions.pristine('success')
        //actions.clear('success')
        actions.error('server')
        res('done')
      }, 3000)
    })
    console.log(result)
  }
  return (
    <Form
      onSubmit={(e, form, actions) => handleSubmit(e, form, actions)}
      stateStrs={stateStrs}
    >
      <div>
        <Input
          name="name"
          label="Your Name"
          required
          match="^Jeff"
          messages={nameMessages}
        />
      </div>
      <div>
        <Input name="age" label="Your Age" min="16" max="65" />
      </div>

      <VisibilityGroup
        isVisible={form => form.fields.age && form.fields.age.value > 18}
      >
        <Input name="numberJobs" label="Number of jobs..." />
        <VisibilityGroup
          isVisible={form =>
            form.fields.numberJobs && form.fields.numberJobs.value > 5
          }
        >
          Wow that is a lot!
          <Input
            name="managerRating"
            label="What would the average manager give on you a scale 1-10?"
          />
        </VisibilityGroup>
      </VisibilityGroup>
      <Input
        name="references[0].name"
        label="Your Reference Name #1"
        required
      />
      <Input name="references[0].phone" label="Phone" required />
      <Input
        name="references[1].name"
        label="Your Reference Name #2"
        required
      />
      <Input name="references[1].phone" label="Phone" required />

      <Input name="skills[0].first" label="Your Best Skill" required />
      <Input name="skills.[1].second" label="Your Second Skill" required />

      <h3>Your Employment History</h3>
      <div>
        <List minLength="2" maxLength="5" name="employers">
          <h5>Employer Entry</h5>
          <RemoveListItem from text="Remove Employer" />
          <div>
            <Input name="employer" label="Employer" required />
            <Input name="title" label="Your Title" required />
            <Input name="duration" label="Years / months duration" />
          </div>
          <h6>Your Managers</h6>
          <List name="managers" minLength="1" maxLength="2">
            <div>
              <Input name="name" label="Manager" />
            </div>
            <RemoveListItem from text="Remove Manager" />
          </List>
          <br />
          <br />
          <AddListItem to="managers" text="Add Manager" />
          <hr />
        </List>
        <div>
          <br />
          <AddListItem to="employers" text="Add Employer" />
        </div>
      </div>

      <div>
        <br />
        <Button text="Submit" type="submit" />
        <SimpleStates />
      </div>
    </Form>
  )
}

export default PList
