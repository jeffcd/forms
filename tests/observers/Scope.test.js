import React from 'react'
import Form from '../../src'
import asField from '../../src/hoc/asField'
import withListActions from '../../src/hoc/withListActions'
import withFormState from '../../src/hoc/withFormState'
import List from '../../src/fields/List'
import Scope from '../../src/observers/Scope'

const Button = withFormState(({ id, text, type, onClick, form }) => {
  const disabled = form.state === 'submit'

  return (
    <button id={id} name={id} type={type} disabled={disabled} onClick={onClick}>
      {text}
    </button>
  )
})

const Input = asField(
  ({
    onChange,
    onBlur = () => {},
    name,
    id,
    value,
    errors,
    label,
    helperText,
    required = false
  }) => {
    return (
      <div>
        <label htmlFor={name}>{label}</label>
        <br />
        <input
          id={`input-${id}`}
          name={name}
          value={value}
          onChange={e => onChange(e)}
          onBlur={e => onBlur(e)}
          label={label}
          required={required}
        />
        <span>{errors.length ? errors[0] : helperText}</span>
      </div>
    )
  }
)

const AddListItem = withListActions(({ id, text, to, listActions }) => {
  return (
    <Button
      id={`add-${to}`}
      text={text}
      type="button"
      onClick={() => listActions.addItemToList(to)}
    />
  )
})

it('Form: Scope with List', () => {
  const submitSpy = jest.fn()
  let bodySpy = {}
  const form = (
    <Form
      onSubmit={(e, body, actions) => {
        bodySpy = body
        submitSpy()
      }}
    >
      <Scope path="users">
        <List name="games" minLength="1">
          <Input name="name" />
        </List>
        <AddListItem to="games" text="Add Game" />
        <Input id="firstName" name="firstName" required />
        <Scope path="skills">
          <Input id="first" name="first" />
          <Input id="second" name="second" />
        </Scope>
      </Scope>
      <button type="submit" id="submit-button">
        Submit
      </button>
    </Form>
  )
  const tree = mount(form)

  expect(tree.html()).toMatchSnapshot()

  tree.find('form').simulate('submit')
  tree.find('#submit-button').simulate('click')
  expect(tree.html()).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(0)

  tree
    .find('#input-firstName')
    .simulate('change', { target: { name: 'firstName', value: 'Jeff' } })
  tree.find('#submit-button').simulate('click')
  tree.find('form').simulate('submit')
  expect(tree.html()).toMatchSnapshot()
  expect(bodySpy).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(1)
})

it('Form: Scope with List', () => {
  const submitSpy = jest.fn()
  let bodySpy = {}
  const form = (
    <Form
      onSubmit={(e, body, actions) => {
        bodySpy = body
        submitSpy()
      }}
    >
      <Scope path="users">
        <Input id="firstName" name="firstName" required />
        <Scope path="skills">
          <Input id="first" name="first" />
          <Input id="second" name="second" />
        </Scope>
      </Scope>
      <Scope>
        <Input id="game" />
      </Scope>
      <button type="submit" id="submit-button">
        Submit
      </button>
    </Form>
  )
  const tree = mount(form)

  expect(tree.html()).toMatchSnapshot()

  tree.find('form').simulate('submit')
  tree.find('#submit-button').simulate('click')
  expect(tree.html()).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(0)

  tree
    .find('#input-firstName')
    .simulate('change', { target: { name: 'firstName', value: 'Jeff' } })
  tree.find('#submit-button').simulate('click')
  tree.find('form').simulate('submit')
  expect(tree.html()).toMatchSnapshot()
  expect(bodySpy).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(1)
})
