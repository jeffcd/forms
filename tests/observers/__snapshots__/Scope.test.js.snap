// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form: Scope with List 1`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"users.games.value[0].name\\"></label><br><input id=\\"input-users.games:0::name:\\" name=\\"users.games.value[0].name\\" value=\\"\\"><span></span></div><button id=\\"add-users.games\\" name=\\"add-users.games\\" type=\\"button\\">Add Game</button><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"first\\"></label><br><input id=\\"input-first\\" name=\\"first\\" value=\\"\\"><span></span></div><div><label for=\\"second\\"></label><br><input id=\\"input-second\\" name=\\"second\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Scope with List 2`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"users.games.value[0].name\\"></label><br><input id=\\"input-users.games:0::name:\\" name=\\"users.games.value[0].name\\" value=\\"\\"><span></span></div><button id=\\"add-users.games\\" name=\\"add-users.games\\" type=\\"button\\">Add Game</button><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span>undefined is required.</span></div><div><label for=\\"first\\"></label><br><input id=\\"input-first\\" name=\\"first\\" value=\\"\\"><span></span></div><div><label for=\\"second\\"></label><br><input id=\\"input-second\\" name=\\"second\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Scope with List 3`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"users.games.value[0].name\\"></label><br><input id=\\"input-users.games:0::name:\\" name=\\"users.games.value[0].name\\" value=\\"\\"><span></span></div><button id=\\"add-users.games\\" name=\\"add-users.games\\" type=\\"button\\" disabled=\\"\\">Add Game</button><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"Jeff\\"><span></span></div><div><label for=\\"first\\"></label><br><input id=\\"input-first\\" name=\\"first\\" value=\\"\\"><span></span></div><div><label for=\\"second\\"></label><br><input id=\\"input-second\\" name=\\"second\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Scope with List 4`] = `
Object {
  "users": Object {
    "firstName": "Jeff",
    "games": Array [
      Object {
        "name": "",
      },
    ],
    "skills": Object {
      "first": "",
      "second": "",
    },
  },
}
`;

exports[`Form: Scope with List 5`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"first\\"></label><br><input id=\\"input-first\\" name=\\"first\\" value=\\"\\"><span></span></div><div><label for=\\"second\\"></label><br><input id=\\"input-second\\" name=\\"second\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Scope with List 6`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span>undefined is required.</span></div><div><label for=\\"first\\"></label><br><input id=\\"input-first\\" name=\\"first\\" value=\\"\\"><span></span></div><div><label for=\\"second\\"></label><br><input id=\\"input-second\\" name=\\"second\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Scope with List 7`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"Jeff\\"><span></span></div><div><label for=\\"first\\"></label><br><input id=\\"input-first\\" name=\\"first\\" value=\\"\\"><span></span></div><div><label for=\\"second\\"></label><br><input id=\\"input-second\\" name=\\"second\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Scope with List 8`] = `
Object {
  "users": Object {
    "firstName": "Jeff",
    "skills": Object {
      "first": "",
      "second": "",
    },
  },
}
`;
